### PR TITLE
Add GRPO into experimental

### PR DESCRIPTION
**Pull request by A9isha**
# Description

Add GRPO into MaxText as experimental code. We follow the implementation from TRL.

This work was done in collaboration by @A9isha and @khatwanimohit 

TODO:
1. The TFLOPS calculation needs to be updated to include the FLOPS in the inference used for autoregressively generating the training data
2. Improve the KL divergence integration test because currently we are using step 0 in the test, where reference params and policy params are identical
3. Currently only tested with a small model (Gemma2-2B) which fits in memory

FIXES: b/395900884
Tested locally on a v5p-8 with following command

`python -m MaxText.experimental.rl.grpo_trainer.py MaxText/experimental/rl/grpo.yml per_device_batch_size=1 tokenizer_path='google/gemma-2-2b-it' run_name=${USER}-$(date +%Y-%m-%d-%H-%M) model_name=gemma2-2b base_output_directory=gs://runner-maxtext-logs dataset_type=hf hf_path='trl-lib/tldr' enable_checkpointing=false`

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/experimental/rl/grpo_trainer.py [hunk 1]
```diff
@@ -0,0 +1,907 @@
...
+  # prompt_segmentation = data["prompt_segmentation"]
+  prompt_completions_position = data["prompt_completions_position"]
+  prompt_completions_segmentation = data["prompt_completions_segmentation"]
+  completions_segmentation = data["ar_completions_segmentation"]
+
+  # compute_log_probs returns logits.
+  # We compute the log-probabilities for the entire generated sequence, then shift as usual.
+  # 'completions' contains only the AR generated tokens, so token_logps_ref and token_logps_policy only contain the logprob for the ar tokens and not the prompts
+  rng1, rng_fwd = random.split(dropout_rng)
+  token_logps_policy, intermediate_outputs = compute_log_probs(
```
Comment (gagika) [inline MaxText/experimental/rl/grpo_trainer.py:155]: perhaps jit-ing for performance, here and below.

File: MaxText/experimental/rl/grpo_trainer.py [hunk 2]
```diff
@@ -0,0 +1,907 @@
...
+  """Complete input prompts"""
+  for k, v in data.items():
+    if v.ndim == 2:
+      data[k] = v[: config.micro_batch_size_to_train_on, :]
+    else:
+      data[k] = v[: config.micro_batch_size_to_train_on]
+
+  rng, rng_gen = random.split(rng)
+  L_prompt = data["prompt_true_length"]
+  data = generate_completions(
```
Comment (gagika) [inline MaxText/experimental/rl/grpo_trainer.py:338]: jitting?

File: MaxText/experimental/rl/grpo_trainer.py [hunk 3]
```diff
@@ -0,0 +1,907 @@
...
+      params,
+      inputs,
+      inputs_position,
+      decoder_segment_ids=inputs_segmentation,
+      enable_dropout=(config.enable_dropout if is_train else False),
+      rngs=rngs,
+      mutable="intermediates",
+  )  # [B, S, E] - [batch, sequence, embedding/vocab]
+  if not is_train:
+    logits = jax.lax.stop_gradient(logits)
```
Comment (gagika) [inline MaxText/experimental/rl/grpo_trainer.py:400]: I think it still keeps activation (uses extra memory).

you could do something like this at the beginning:

params = params if training  else jax.lax.stop_gradient(params)

File: assets/gemma2-2b/tokenizer.json
Comment (gagika) [inline assets/gemma2-2b/tokenizer.json:1]: why do you need a new tokenizer?

File: MaxText/experimental/rl/grpo_input_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    # def _get_local_device_buffers(arr):
+    #   try:
+    #     local_device_arrays = np.split(arr, len(self.global_mesh.local_devices), axis=0)
+    #   except ValueError as array_split_error:
+    #     raise ValueError(
+    #         f"Unable to put to devices shape {arr.shape} with "
+    #         f"local device count {len(self.global_mesh.local_devices)} "
+    #     ) from array_split_error
+    #   return jnp.vstack(jax.device_put(local_device_arrays, self.global_mesh.local_devices))
+    # local_device_buffers = jtu.tree_map(partial(_get_local_device_buffers), local_data)
```
[the remark this answers is not in the document]
Comment (khatwanimohit) [reply MaxText/experimental/rl/grpo_input_pipeline.py]: Removed

File: MaxText/experimental/rl/grpo_trainer.py [hunk 5]
```diff
@@ -0,0 +1,907 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+# pylint: disable=g-bad-todo, abstract-method, consider-using-with, ungrouped-imports
+"""Training loop and Decoding of the model."""
```
[the remark this answers is not in the document]
Comment (khatwanimohit) [reply MaxText/experimental/rl/grpo_trainer.py]: Done

File: MaxText/experimental/rl/grpo_trainer.py (hunk 2 above)
Comment (khatwanimohit) [reply MaxText/experimental/rl/grpo_trainer.py]: Done

File: MaxText/experimental/rl/grpo_trainer.py (hunk 1 above)
Comment (khatwanimohit) [reply MaxText/experimental/rl/grpo_trainer.py:154]: This function is called inside loss function which is called by train_step which is jitted. Do we still need to jit this ?

File: MaxText/experimental/rl/grpo_trainer.py (hunk 3 above)
Comment (khatwanimohit) [reply MaxText/experimental/rl/grpo_trainer.py]: Done

File: assets/gemma2-2b/tokenizer.json
Comment (khatwanimohit) [reply assets/gemma2-2b/tokenizer.json:1]: Removed the tokenizer

File: MaxText/experimental/rl/decode_grpo.py
Comment (khatwanimohit) [inline MaxText/experimental/rl/decode_grpo.py:1]: @A9isha Do we need this file? we could add support for using HF tokenizer in our decode.py instead

Comment (khatwanimohit) [reply MaxText/experimental/rl/decode_grpo.py:1]: I have removed decode_grpo.py as the feature of loading HF tokenizer in decode.py is going to come in this #1439

Comment (A9isha) [reply MaxText/experimental/rl/decode_grpo.py:1]: thanks Mohit!

File: MaxText/tests/integration_tests/grpo_correctness.py [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+        reward_funcs=reward_len,
+        args=training_args,
+        train_dataset=load_dataset("trl-lib/tldr", split="train"),
+    )
+    self.hf_model = transformers.AutoModelForCausalLM.from_pretrained(
+        "meta-llama/Llama-3.1-8B",
+        torch_dtype=torch.float32,
+    )
+
+  # def test_inference(self):
```
Comment (gagika) [inline MaxText/tests/integration_tests/grpo_correctness.py:113]: Do you need this commented code?

File: MaxText/experimental/rl/grpo.yml [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+enable_goodput_recording: False
+monitor_goodput: False
+dataset_type: hf 
+hf_path: 'json' 
+hf_data_dir: '' 
+
+#TRL
+max_prefill_predict_length: 512
+max_target_length: 1024
+hf_train_files: 'gs://mazumdera-test-bucket-europe-west4/rl-datasets/tldr_dataset.jsonl'
```
Comment (gagika) [inline MaxText/experimental/rl/grpo.yml:20]: Should it be part of base grpo.yml ?

Let's pass it during each run.

File: MaxText/experimental/rl/grpo.yml [hunk 8]
```diff
@@ -0,0 +1,33 @@
+base_config: "base.yml"
+
+use_grpo: True
+train_data_columns: ['prompt', 'completion']
+tokenizer_path: 'google/gemma-2-2b-it' 
+model_name: gemma2-2b
```
Comment (gagika) [inline MaxText/experimental/rl/grpo.yml:6]: Same as below, do we need to fix a specific model/tokenizer in the base config?

**Review by gagika — APPROVED** [review]
Only a few comments

File: MaxText/tests/integration_tests/grpo_correctness.py (hunk 6 above)
Comment (khatwanimohit) [reply MaxText/tests/integration_tests/grpo_correctness.py]: removed

File: MaxText/experimental/rl/grpo.yml (hunk 7 above)
Comment (khatwanimohit) [reply MaxText/experimental/rl/grpo.yml]: removed

File: MaxText/experimental/rl/grpo.yml (hunk 8 above)
Comment (khatwanimohit) [reply MaxText/experimental/rl/grpo.yml]: removed

**Review by khatwanimohit — APPROVED** [review]

File: MaxText/experimental/rl/grpo_trainer.py [hunk 9]
```diff
@@ -0,0 +1,883 @@
...
+  rewards = jnp.array(rewards)  # shape [BxG]
+
+  # --- (7) Group rewards and compute normalized advantage.
+  G = config.num_generations
+  rewards_grouped = rewards.reshape(-1, G)  # shape [B, G]
+  group_mean = jnp.mean(rewards_grouped, axis=1)  # shape [B]
+  group_std = jnp.std(rewards_grouped, axis=1)  # shape [B]
+  repeated_group_mean = jnp.repeat(group_mean, G)  # shape [BxG]
+  repeated_group_std = jnp.repeat(group_std, G)  # shape [BxG]
+  advantages = (rewards - repeated_group_mean) / (repeated_group_std + 1e-4)  # shape [BxG]
```
Comment (richjames0) [inline MaxText/experimental/rl/grpo_trainer.py:200]: nit hoist 1e4 as a constant

File: MaxText/experimental/rl/grpo_trainer.py [hunk 10]
```diff
@@ -0,0 +1,883 @@
...
+    total_moe_lb_loss = maxtext_utils.get_nested_value(intermediate_outputs, nested_key, 0.0)
+    moe_lb_loss = jnp.mean(jnp.array(total_moe_lb_loss))
+    loss += moe_lb_loss
+
+  # --- (9) Compute auxiliary metrics.
+  avg_kl = jnp.mean((per_token_kl * valid_seq_mask) / (jnp.sum(valid_seq_mask, axis=1, keepdims=True) + EPS))
+  avg_reward = jnp.mean(rewards)
+  avg_advantage = jnp.mean(advantages)
+  avg_completion_length = jnp.mean(jnp.sum(data["ar_completions_segmentation"] != 0, axis=1))
+  aux = {
```
Comment (richjames0) [inline MaxText/experimental/rl/grpo_trainer.py:229]: IMO this would be better as a dataclass

File: MaxText/experimental/rl/grpo_trainer.py [hunk 11]
```diff
@@ -0,0 +1,883 @@
...
+    tokenizer_model: Tokenizer for generate
+    true_length: Length of the prompt out of the max_target_length
+
+  Returns:
+    A jnp.array of shape [B x num_generations, S] where S = length_of_prompt + max_completion_length.
+  """
+  # decimate proportion of data when per_device_batch_size<1
+  for k, v in data.items():
+    if v.ndim == 2:
+      data[k] = v[: config.micro_batch_size_to_train_on, :]
```
Comment (richjames0) [inline MaxText/experimental/rl/grpo_trainer.py:266]: i believe it's more idiomatic to do :to_num. In fact reading it with current spacing I mis-read it as `:, bsz, :`

File: MaxText/experimental/rl/grpo_trainer.py [hunk 12]
```diff
@@ -0,0 +1,883 @@
...
+    else:
+      data[k] = v[: config.micro_batch_size_to_train_on]
```
Comment (richjames0) [inline MaxText/experimental/rl/grpo_trainer.py:268]: this feels like a good spot for an assertion that v.dim is 1 (or probably better before the if assert it's 1 or 2).

File: MaxText/experimental/rl/grpo_trainer.py [hunk 13]
```diff
@@ -0,0 +1,883 @@
...
+
+  prompts = data["prompt"]
+  rng, rng_load_params = jax.random.split(rng)
+  G = config.num_generations
+
+  rng, rng_init_decode = jax.random.split(rng_load_params)
+  decode_state = engine.init_decode_state(rng_init_decode)
+  slot = 0
+  for i in range(prompts.shape[0]):
```
Comment (richjames0) [inline MaxText/experimental/rl/grpo_trainer.py:277]: for prompt, length in zip(prompts, true_lengths) is more idiomatic

File: MaxText/tests/integration_tests/grpo_correctness.py [hunk 14]
```diff
@@ -0,0 +1,254 @@
+#  Copyright 2025 Google LLC
```
[the remark this answers is not in the document]
Comment (A9isha) [reply MaxText/tests/integration_tests/grpo_correctness.py:1]: Thanks Rich - Added MaxText/tests/grpo_trainer_correctness_test.py - please take a look

**Review by richjames0 — APPROVED** [review]
LGTM!
